### PR TITLE
OGM-887 Don't map elements within embeddable collections using dot names

### PR DESCRIPTION
**Pull request by mp911de**
* Don't map elements within embeddable collections using dot names
* Remove the common prefixes from the JSON representation
* Refactor parts of MongoHelpers into DotPatternMapHelpers

### CODE REVIEW

File: mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/dialect/impl/MongoDBAssociationSnapshot.java [hunk 1]
```diff
@@ -46,11 +47,11 @@ public DBObject getQueryObject() {
 
 		if ( storageStrategy == AssociationStorageStrategy.IN_ENTITY ) {
 			if ( associationKey.getMetadata().isOneToOne() ) {
-				Object oneToOneValue = getValueOrNull( document, associationKey.getMetadata().getCollectionRole(), Object.class );
+				Object oneToOneValue = getValueOrNull( (Map<?,?>) document, associationKey.getMetadata().getCollectionRole(), Object.class );
```
Comment (gunnarmorling) [inline mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/dialect/impl/MongoDBAssociationSnapshot.java]: `DBObject` is not safe to be cast to `Map`. The default implementation `BasicDBObject` indeed is, but it's a slippery slope we'd enter when relying on it. I'd prefer to duplicate that code for now (apparently, MongoDB driver 3 comes with a replacement for `DBObject` et al. anyways, so we may re-visit the topic then).

File: core/src/main/java/org/hibernate/ogm/dialect/impl/DotPatternMapHelpers.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
+/*
+ * Hibernate OGM, Domain model persistence for NoSQL datastores
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.ogm.dialect.impl;
```
Comment (gunnarmorling) [inline core/src/main/java/org/hibernate/ogm/dialect/impl/DotPatternMapHelpers.java]: It'd be better located in `org.hibernate.ogm.datastore.document.impl`.

Btw. to give some background, the "*.spi" packages are meant to provide the required types/functionality for implementing dialects. We give some guarantees regarding compatible evolvement there. Other things where we don't feel like committing to that yet, we have under "impl" and still refer them from "our" dialects. The idea is to promote those things to "spi" if they have proven over time. This helper also feels like it's rightly located under "impl" for now and once we see it fits for different use cases / dialects, we can relocate it in a subsequent release.

**Comment by gunnarmorling** [comment]
@mp911de Looks good to me apart from the two remarks above. Could you also rebase it to the latest master? Thanks a lot!

File: mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/dialect/impl/MongoDBAssociationSnapshot.java (hunk 1 above)
Comment (mp911de) [inline mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/dialect/impl/MongoDBAssociationSnapshot.java]: Ok, will revert the changes to the Mongo code

**Comment by mp911de** [comment]
Thanks for reviewing, I fixed the two issues and rebased the code on the latest master.

**Comment by gunnarmorling** [comment]
Merged; Thanks, @mp911de!
